proposals/0000-no-sugared-incompleteness.rst

```diff
@@ -0,0 +1,211 @@
+But many of us also would say it's a solved problem, because we have ``-Werror=...`` to catch all these things.
+Why write yet another proposal, for yet another language extension, when this is solved problem?
+
+Well, I just don't think ``-Werror=...`` is the right mechanism to rely on.
+This is, of course, a matter of opinion, but to me warnings are for "extra" analyses, not "essential" ones like type checking which enforce the language proper.
+Deciding whether e.g. pattern matching is exhaustive feels to me like one of those essential analyses.
+I am asking that we switch our vantage point and instead of considering programs with implicit partiality well-formed programs that are iffy, consider them ill-formed programs that we can permissively accept anyways.
+
+So, what's the practical upshot of this lawyering?
+And what mechanism might we use rather than ``-Werror=...``?
+There is another mechanism for controlling run-time errors which I think is a perfect model for how these programs should be dealt with: ``-fdefer-type-errors``.
+As I mentioned above, nobody would propose type checking is some extraneous analysis a Haskell implementation need not do by default.
+The "defer" in the name makes clear we are taking what is normally a compile-time error and deferring it to run-time behind additional branching.
+Whether for quick slap-dash programming, or arguing with proponents of untyped languages, ``-fdefer-type-errors`` allows GHC to both express its opinion and yet also not straight-jacket the user into any one policy.
+
+My idea is to do the same thing for exhaustiveness checking: it should be mandatory, but as a debugging aid there are ``-fdefer-...`` flags to defer compile time errors into a run-time errors.
+
+Proposed Change Specification
+-----------------------------
+
+#. Let there be a new language extension ``Incomplete``.
+   This extension is on by default, as that corresponds to the status quo.
+
+   We first describe the behavior of ``NoIncomplete``.
+   This restricts the programs accepted compared to the status quo.
+   More important than the specific specification is the guiding principle behind that specification.
+   It is this: to prohibit any program that can fail for reasons not explicitly specified by the programmer except for non-termination and exhaustion of resources.
+   For GHC, that means to prohibit any program for which GHC would need to emit a synchronous exception throw not specified by the user.
+
+   To accomplish this, we will mainly rely on existing analysis.
+   Any program that would emit warnings from the following warning categories is prohibited:
+
+   - ``incomplete-patterns``
+   - ``incomplete-uni-patterns``
+   - ``incomplete-record-updates``
+   - ``missing-fields``
+   - ``missing-methods``
+
+   Additionally, with ``-XFieldSelectors`` (also on by default), any program that would emit warnings with ``-Wpartial-fields`` is also prohibited.
```

Comment:
Or we could implement `-Wincomplete-record-selectors` and add it to the list above, then `-Wpartial-fields` would be safe? After all, `-Wpartial-fields` doesn't itself introduce incompleteness, so it would be a shame to forbid it if we can be more precise.

Reply:
@adamgundry I totally agree, I just felt it would be cheating specify with reference to an open issue. I would love to tighten this spec up with your new warnings!